### PR TITLE
Implement datasource type aware datasource and planner configuration

### PR DESCRIPTION
**Pull request by devsergiy**
ENG-4945

### CODE REVIEW

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource_federation_test.go [hunk 1]
```diff
@@ -3117,59 +3201,66 @@ func TestGraphQLDataSourceFederation(t *testing.T) {
...
+							FieldNames: []string{"id"},
+						},
+						{
+							TypeName:   "Admin",
+							FieldNames: []string{"adminID", "title"},
 						},
 						{
-							TypeName:     "Admin",
-							SelectionSet: "adminID",
+							TypeName:   "Moderator",
```
Comment (jensneuse) [inline v2/pkg/engine/datasource/graphql_datasource/graphql_datasource_federation_test.go:3229]: Intentional?

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 2]
```diff
@@ -33,83 +34,159 @@ type DataSourceConfiguration struct {
...
+	return d.ChildNodes
+
+}
+
+// dataSourceConfiguration is the configuration for a DataSource
+type dataSourceConfiguration[T any] struct {
+	*DataSourceMetadata                   // DataSourceMetadata is the information about root and child nodes and federation metadata if applicable
+	ID                  string            // ID is a unique identifier for the DataSource
+	Factory             PlannerFactory[T] // Factory is the factory for the creation of the concrete DataSourcePlanner
+	Custom              json.RawMessage
```
Comment (jensneuse) [inline v2/pkg/engine/plan/datasource_configuration.go]: Obsolete?

File: v2/pkg/engine/plan/datasource_configuration.go [hunk 3]
```diff
@@ -33,83 +34,159 @@ type DataSourceConfiguration struct {
...
+	NewCustom           T // NewCustom is the datasource specific configuration
+
+	hash DSHash // hash is a unique hash for the dataSourceConfiguration used to match datasources
```
Comment (jensneuse) [inline v2/pkg/engine/plan/datasource_configuration.go:95]: Please ensure that the hash is created during the configuration stage, not during the planning stage.
Once this is done, we can remove the pre-generation of DSHash at Router startup.

File: v2/pkg/engine/plan/datasource_filter_visitor_test.go [hunk 4]
```diff
@@ -33,6 +35,9 @@ func (b *dsBuilder) ChildNode(typeName string, fieldNames ...string) *dsBuilder
 
 func (b *dsBuilder) Schema(schema string) *dsBuilder {
 	b.ds.Custom = []byte(schema)
+	def := unsafeparser.ParseGraphqlDocumentString(schema)
```
Comment (jensneuse) [inline v2/pkg/engine/plan/datasource_filter_visitor_test.go:37]: line above obsolete?

File: v2/pkg/engine/plan/planner_configuration.go [hunk 5]
```diff
@@ -88,16 +197,24 @@ func (p *plannerConfiguration) fragmentPaths() (out []string) {
...
+func (p *plannerPathsConfiguration) RemovePath(path string) {
+	idx, ok := p.index[path]
+	if !ok {
+		return
 	}
+
+	p.paths[idx] = pathConfiguration{}
+	delete(p.index, path)
+
+	// for i := range p.paths {
```
Comment (jensneuse) [inline v2/pkg/engine/plan/planner_configuration.go]: remove?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource_federation_test.go (hunk 1 above)
Comment (devsergiy) [reply v2/pkg/engine/datasource/graphql_datasource/graphql_datasource_federation_test.go:3229]: config is fully the same, it is just diff, cause lines were moved

**Comment by devsergiy** [comment]
do not merge yet, at the current state it is not possible to pre init factory with context
interface for doing that is not implemented

File: v2/pkg/engine/plan/federation_metadata.go [hunk 6]
```diff
@@ -8,6 +8,29 @@ type FederationMetaData struct {
...
 
+type FederationInfo interface {
+	HasKeyRequirement(typeName, requiresFields string) bool
+	RequiredFieldsByKey(typeName string) []FederationFieldConfiguration
+	RequiredFieldsByRequires(typeName, fieldName string) []FederationFieldConfiguration
+	HasEntity(typeName string) bool
+}
+
+func (d *FederationMetaData) HasKeyRequirement(typeName, requiresFields string) bool {
+	return d.Keys.HasSelectionSet(typeName, "", requiresFields)
```
Comment (jensneuse) [inline v2/pkg/engine/plan/federation_metadata.go:19]: Is empty string correct?

File: v2/pkg/engine/plan/planner.go [hunk 7]
```diff
@@ -313,8 +316,8 @@ func (p *Planner) printPlanningPaths() {
 	p.debugMessage("Planning paths:")
 	for i, planner := range p.configurationVisitor.planners {
 		fmt.Println("Paths for planner", i+1)
-		fmt.Println("Planner parent path", planner.parentPath)
-		for _, path := range planner.paths {
+		fmt.Println("Planner parent path", planner.ParentPath())
```
Comment (jensneuse) [inline v2/pkg/engine/plan/planner.go:319]: this Println is only called in debug mode, right?

File: v2/pkg/graphql/config_factory_federation.go [hunk 8]
```diff
@@ -53,7 +54,12 @@ func WithFederationSubscriptionType(subscriptionType SubscriptionType) Federatio
 	}
 }
 
-func NewFederationEngineConfigFactory(dataSourceConfigs []graphqlDataSource.Configuration, opts ...FederationEngineConfigFactoryOption) *FederationEngineConfigFactory {
+type DataSourceConfiguration struct {
+	ID            string
```
Comment (jensneuse) [inline v2/pkg/graphql/config_factory_federation.go]: Can you document the meaning of the fields?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: v2/pkg/engine/plan/federation_metadata.go (hunk 6 above)
Comment (devsergiy) [reply v2/pkg/engine/plan/federation_metadata.go:19]: yes, keys do not have field name

File: v2/pkg/engine/plan/planner.go (hunk 7 above)
Comment (devsergiy) [reply v2/pkg/engine/plan/planner.go:319]: yeah, it is called like this
```
	if p.config.Debug.PrintPlanningPaths {
		p.printPlanningPaths()
	}
```

**Review by jensneuse — APPROVED** [review]

**Pull request merged**